### PR TITLE
Add invalid withdrawal detection to bots.

### PR DESCRIPTION
**Pull request by chrismaree**
This PR adds invalid withdrawal detection to the liquidator and dispute bots by modifying the `ExpiringMultiPartyClient`'s `_isUnderCollateralized` function.

Previously a positions collateralization did not consider any pending withdrawals. After this PR a positions collateralization is considered their `amountCollateral` net any withdrawals as:

```
(numTokens * trv) * collateralRequirement > collateralNetWithdrawal.
```

### CODE REVIEW

File: disputer/test/Disputer.js [hunk 1]
```diff
@@ -45,7 +45,7 @@ contract("Disputer.js", function(accounts) {
   const unreachableDeadline = MAX_UINT_VAL;
 
   before(async function() {
-    collateralToken = await Token.new("UMA", "UMA", 18, { from: contractCreator });
+    collateralToken = await Token.new("DAI", "DAI", 18, { from: contractCreator });
```
Comment (chrismaree) [inline disputer/test/Disputer.js:48]: unrelated change: collateral should not be called `UMA`

File: disputer/test/Disputer.js [hunk 2]
```diff
@@ -56,17 +56,17 @@ contract("Disputer.js", function(accounts) {
     await collateralToken.mint(sponsor3, toWei("100000"), { from: contractCreator });
     await collateralToken.mint(liquidator, toWei("100000"), { from: contractCreator });
     await collateralToken.mint(disputeBot, toWei("100000"), { from: contractCreator });
+  });
```
Comment (chrismaree) [inline disputer/test/Disputer.js:59]: moved the creation of the `mockOracle` to the `beforeEach` from the `before` to decouple the tests.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 3]
```diff
@@ -142,17 +147,18 @@ class ExpiringMultiPartyClient {
       lastUpdateTimestamp: this.lastUpdateTimestamp
     });
   };
-  _isUnderCollateralized = (numTokens, amountCollateral, trv) => {
+  _isUnderCollateralized = (numTokens, amountCollateral, withdrawalRequest, trv) => {
```
Comment (chrismaree) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:150]: this is the key change of the PR: adding this `withdrawalRequest` parameter.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 4]
```diff
@@ -51,7 +56,7 @@ class ExpiringMultiPartyClient {
   // Whether the given undisputed `liquidation` (`getUndisputedLiquidations` returns an array of `liquidation`s) is disputable.
   // `tokenRedemptionValue` should be the redemption value at `liquidation.time`.
   isDisputable = (liquidation, tokenRedemptionValue) => {
-    return !this._isUnderCollateralized(liquidation.numTokens, liquidation.amountCollateral, tokenRedemptionValue);
+    return !this._isUnderCollateralized(liquidation.numTokens, liquidation.amountCollateral, "0", tokenRedemptionValue);
```
Comment (chrismaree) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:59]: one downside with adding the parameter to the `_isUnderCollateralized` function is that the `withdrawalRequest` is always zero for a liquidation. An alternative implementation would be to make the call to `this._isUnderCollateralized` on line 34 be this instead:

```
this._isUnderCollateralized(
        position.numTokens,
        toBN(position.amountCollateral).sub(toBN(position.withdrawalRequestAmount)).toString(),
        tokenRedemptionValue
      )
```

This has the same effect but does not change the `_isUnderCollateralized` function. Thoughts on these options?

File: disputer/test/Disputer.js [hunk 5]
```diff
@@ -183,6 +183,50 @@ contract("Disputer.js", function(accounts) {
     assert.equal((await emp.getLiquidations(sponsor3))[0].disputer, disputeBot);
   });
 
+  it("Detect disputable withdraws and send dipsutes", async function() {
```
Comment (nicholaspai) [inline disputer/test/Disputer.js:186]: typo "disputes"

File: liquidator/test/Liquidator.js [hunk 6]
```diff
@@ -187,6 +187,73 @@ contract("Liquidator.js", function(accounts) {
...
+
+    // Advance the timer to the liquidation expiry.
+    const liquidationTime = liquidationObject.liquidationTime;
+    const liquidationLiveness = 1000;
+    await emp.setCurrentTime(Number(liquidationTime) + liquidationLiveness);
+
+    // Now that the liquidation has expired, the liquidator can withdraw rewards.
+    const collateralPreWithdraw = await collateralToken.balanceOf(liquidatorBot);
+    await liquidator.queryAndWithdrawRewards();
+    assert.equal(spy.callCount, 2); // 1 info level events should be sent at the conclusion of the withdrawal.
```
Comment (nicholaspai) [inline liquidator/test/Liquidator.js:239]: This comment should say 2 info level events

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js (hunk 4 above)
Comment (nicholaspai) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js:59]: +1 this is actually very confusing. I think your suggested approach is better because it makes more sense for `_isUnderCollateralized` to return simply if the a given collateral ratio is too low. Whether that collateral include requested withdrawal amounts or not should be left to the caller. Also, if you need to pass in a "0" in many cases, then you might as well make that change upstream.

Nevertheless, I would add a comment in `isDisputable` that the way `liquidation.amountOfCollateral` is computed in the first place is "net of requested withdrawal amounts". This way the user knows that the collateral in question includes requested withdrawals.

File: liquidator/test/Liquidator.js (hunk 6 above)
Comment (chrismaree) [reply liquidator/test/Liquidator.js:239]: nope. the comment is correct. the first info level message was sent on line 219. this is the second one. such that the total is equal to 2. The comment could be more descriptive though to make it clear that it should be 1+1 info level events.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js (hunk 4 above)
Comment (chrismaree) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js:59]: ok cool. I'll implement the change.

File: disputer/test/Disputer.js (hunk 1 above)
Comment (nicholaspai) [reply disputer/test/Disputer.js:48]: _everything_ should be called `UMA`

File: disputer/test/Disputer.js (hunk 5 above)
Comment (nicholaspai) [inline disputer/test/Disputer.js:186]: You still missed this comment

**Comment by chrismaree** [comment]
@mrice32 would love some input so I can get this merged in tomorrow.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js (hunk 4 above)
Comment (mrice32) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js]: Agreed. I like the second approach better.

File: financial-templates-lib/clients/ExpiringMultiPartyClient.js [hunk 7]
```diff
@@ -28,11 +28,18 @@ class ExpiringMultiPartyClient {
...
+  // `_isUnderCollateralized` is taken as the positions `amountCollateral` minus any `withdrawalRequestAmount`. As a
+  // result this function will return positions that are undercollateralized due to too little collateral or a withdrawal
+  // that, if passed, would make the position undercollateralized.
   getUnderCollateralizedPositions = tokenRedemptionValue => {
-    return this.positions.filter(position =>
-      this._isUnderCollateralized(position.numTokens, position.amountCollateral, tokenRedemptionValue)
-    );
+    return this.positions.filter(position => {
+      const collateralNetWithdrawal = this.web3.utils
+        .toBN(position.amountCollateral)
```
Comment (mrice32) [inline financial-templates-lib/clients/ExpiringMultiPartyClient.js:38]: nit: going forward, aren't we supposed to be declaring these functions at the top so we can get rid of the `web3.utils`? Note: it's cool with me if you just change the new code and leave other code in the same file using the `web3.utils` syntax.

Comment (chrismaree) [reply financial-templates-lib/clients/ExpiringMultiPartyClient.js:38]: ye I didnt want to make this change in this PR ad we can leave it for one larger chance in issue [1411](https://github.com/UMAprotocol/protocol/issues/1411). Also, this one is a bit more complex because there is no scoped web3 instance so we'll leave it for later.
